File: 2023/02/2.rb
#!/usr/bin/env ruby
part_1_max_cubes = { 'red' => 12, 'green' => 13, 'blue' => 14 }
results = $stdin.readlines.each_with_index.map do |line, game_id|
  max_cubes = Hash.new(0)
  line.strip.split(':').last.strip.split(/\s*[;,]\s*/).each do |cube|
    number, color = cube.split(' ')
    max_cubes[color] = [max_cubes[color], number.to_i].max
  end
  game_id = -1 unless max_cubes.all? { |color, count| part_1_max_cubes[color] >= count }
  [game_id + 1, max_cubes.values.inject(1, &:*)]
end
sums = results.inject([0, 0]) do |(part1, part2), (game_id, product)|
  [part1 + game_id, part2 + product]
end
puts(*sums)

File: 2023/02/2.go
package main

import (
	"bufio"
	"fmt"
	"regexp"
	"strings"
	"strconv"
	"os"
)

func main() {
  scanner := bufio.NewScanner(os.Stdin)

  re := regexp.MustCompile(`\d+ \w+`)

  part_1_thresholds := map[string]int {"red": 12,"green": 13,"blue": 14}
  game_id, part1_sum, part2_sum := 1, 0, 0

  for scanner.Scan() {
  	cubes := re.FindAllString(strings.Split(scanner.Text(), ":")[1], -1)
	  max_cubes := map[string]int {"red": 0,"green": 0,"blue": 0}

  	for _, cube := range(cubes) {
  		cube_parts := strings.Split(cube, " ")
  		number, _ := strconv.Atoi(cube_parts[0])
  		color := cube_parts[1]
  		if number > max_cubes[color] {
  			max_cubes[color] = number
  		}
  	}

  	part1_threshold := true
  	p := 1
  	for c, n := range(max_cubes) {
  		p = p * n
  		if part_1_thresholds[c] < n {
  			part1_threshold = false
  		}
  	}

    if part1_threshold {
    	part1_sum = part1_sum + game_id
    }
    part2_sum = part2_sum + p
  	game_id = game_id + 1
  }

  fmt.Println(part1_sum, part2_sum)
}
